src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,363 @@
+    -- (non-zero) overflow pages. This entry can only come from the input run.
+  | OverflowEntry
+      !(Entry SerialisedValue (BlobRef (Run fhandle)))
+      !RawPage
+      !Word32
+      ![RawOverflowPage]
+
+prefixEntryInvariant :: PrefixEntry fhandle -> Bool
+prefixEntryInvariant FullEntry{} = True
+prefixEntryInvariant (OverflowEntry e page lenSuffix overflowPages)
+  | Mupdate _ <- e = False
+  | otherwise =
+       not (null overflowPages)
+    && rawPageOverflowPages page == ceilDivPageSize (fromIntegral lenSuffix)
+    && rawPageOverflowPages page == length overflowPages
+
+toFullEntry :: PrefixEntry fhandle -> Entry SerialisedValue (BlobRef (Run fhandle))
+toFullEntry (FullEntry e) =
+    e
+toFullEntry (OverflowEntry e _ lenSuffix overflowPages) =
+    first appendSuffix e
+  where
+    appendSuffix (SerialisedValue prefix) =
+      SerialisedValue $ RB.take (RB.size prefix + fromIntegral lenSuffix) $
+        mconcat (prefix : map rawOverflowPageRawBytes overflowPages)
+
+-- | Returns 'Nothing' if no input 'Run' contains any entries.
+-- The list of runs should be sorted from new to old.
+new ::
+     HasFS IO h
+  -> HasBufFS IO h
+  -> Level
+  -> Mappend
+  -> Run.RunFsPaths
+  -> [Run (FS.Handle h)]
+  -> IO (Maybe (Merge (FS.Handle h)))
+new fs bfs mergeLevel mergeMappend targetPaths runs = do
+    mreaders <- newReaders fs bfs runs
+    for mreaders $ \(mergeReaders, firstKey, firstEntry) -> do
+      mergeCurrentKey <- newIORef firstKey
+      mergeCurrentEntry <- newIORef firstEntry
+
+      -- calculate upper bounds based on input runs
+      let numEntries = coerce (sum @[] @Int) (map Run.runNumEntries runs)
+      let numPages = sum (map Run.sizeInPages runs)
+      mergeBuilder <- Builder.new fs targetPaths numEntries numPages
+
+      return Merge {..}
+
+data StepResult fhandle = MergeInProgress | MergeComplete !(Run fhandle)
+
+-- | Do a fixed number of steps of merging. Each step reads a single entry, then
+-- either resolves it with the previous one or writes the previous one out to
+-- the run being created.
+--
+-- The resulting run has a reference count of 1.
+--
+-- Do not call again after 'MergeComplete' has been returned!
+steps ::
+     HasFS IO h
+  -> HasBufFS IO h
+  -> Merge (FS.Handle h)
+  -> Int
+  -> IO (StepResult (FS.Handle h))
+steps fs bfs Merge {..} = \numSteps -> do
+    curKey <- readIORef mergeCurrentKey
+    curEntry <- readIORef mergeCurrentEntry
+    go curKey curEntry numSteps
+  where
+    go !curKey !curEntry n
+      | n <= 0 = do
+          writeIORef mergeCurrentKey curKey
+          writeIORef mergeCurrentEntry curEntry
+          return MergeInProgress
+    go !curKey !curEntry !n =
+      getFromReaders fs bfs mergeReaders >>= \case
+        Just (newKey, newEntry) | newKey == curKey -> do
+          -- still same key: don't write yet, resolve!
+          let resolvedEntry = resolveEntries mergeMappend curEntry newEntry
+          go curKey resolvedEntry (n-1)
```

Comment:
This could probably be optimised in the normal case, since we can skip over all elements that have the same key

Reply:
Yes, that's one of the things that are more straight-forward in the other approach (read multiple, write one entry). However, we still have to read the disk pages to know whether the key matches, so the only thing we can really avoid is parsing out the entry and potentially reading overflow pages. It might even be possible to use the compact index to avoid reading some entries entirely, but I don't think I want to get into these complications for now.